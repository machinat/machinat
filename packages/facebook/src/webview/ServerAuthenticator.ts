import { serviceProviderClass } from '@sociably/core/service';
import {
  ServerAuthenticator,
  VerifyResult,
  CheckDataResult,
} from '@sociably/auth';
import BasicAuthenticator from '@sociably/auth/basicAuth';
import { MetaApiError } from '@sociably/meta-api';
import { attachWebviewParamsOnUrl } from '@sociably/webview/client';
import BotP from '../Bot.js';
import ProfilerP from '../Profiler.js';
import FacebookPage from '../Page.js';
import FacebookChat from '../Chat.js';
import FacebookUser from '../User.js';
import { FACEBOOK } from '../constant.js';
import { AgentSettingsAccessorI } from '../interface.js';
import { getAuthContextDetails } from './utils.js';
import type { FacebookAuthContext, FacebookAuthData } from './types.js';

/** @category Provider */
export class FacebookServerAuthenticator
  implements ServerAuthenticator<never, FacebookAuthData, FacebookAuthContext>
{
  private profiler: ProfilerP;
  basicAuthenticator: BasicAuthenticator;
  settingsAccessor: AgentSettingsAccessorI;
  delegateAuthRequest: ServerAuthenticator<
    never,
    FacebookAuthData,
    FacebookAuthContext
  >['delegateAuthRequest'];

  platform = FACEBOOK;

  constructor(
    bot: BotP,
    profiler: ProfilerP,
    basicAuthenticator: BasicAuthenticator,
    settingsAccessor: AgentSettingsAccessorI,
  ) {
    this.profiler = profiler;
    this.basicAuthenticator = basicAuthenticator;
    this.settingsAccessor = settingsAccessor;
    this.delegateAuthRequest = this.basicAuthenticator.createRequestDelegator<
      FacebookAuthData,
      FacebookAuthData,
      FacebookChat
    >({
      bot,
      platform: FACEBOOK,
      platformName: 'Facebook',
      platformColor: '#4B69FF',
      platformImageUrl: 'https://sociably.js.org/img/icon/messenger.png',
      checkCurrentAuthUsability: (credential, data) => ({
        ok: credential.page === data.page && credential.user === data.user,
      }),
      verifyCredential: async (credential) => {
        const { page: pageId, user: userId } = credential;
        return this._verifyUser(pageId, userId);
      },
      checkAuthData: (data) => {
        const result = this.checkAuthData(data);
        if (!result.ok) {
          return result;
        }
        return {
          ok: true,
          data,
          thread: result.contextDetails.thread,
          chatLinkUrl: `https://m.me/${data.page}`,
        };
      },
    });
  }

  getAuthUrl(
    user: FacebookUser,
    options?: {
      redirectUrl?: string;
      webviewParams?: Record<string, unknown>;
    },
  ): string {
    const authUrl = this.basicAuthenticator.getAuthUrl<FacebookAuthData>(
      FACEBOOK,
      { page: user.pageId, user: user.id },
      options?.redirectUrl,
    );
    return options?.webviewParams
      ? attachWebviewParamsOnUrl(authUrl, options.webviewParams)
      : authUrl;
  }

  // eslint-disable-next-line class-methods-use-this
  async verifyCredential(): Promise<VerifyResult<FacebookAuthData>> {
    return {
      ok: false as const,
      code: 403,
      reason: 'should use backend based flow only',
    };
  }

  async verifyRefreshment(
    data: FacebookAuthData,
  ): Promise<VerifyResult<FacebookAuthData>> {
    return this._verifyUser(data.page, data.user);
  }

  // eslint-disable-next-line class-methods-use-this
  checkAuthData(data: FacebookAuthData): CheckDataResult<FacebookAuthContext> {
    return {
      ok: true,
      contextDetails: getAuthContextDetails(data),
    };
  }

  private async _verifyUser(
    pageId: string,
    userId: string,
  ): Promise<VerifyResult<FacebookAuthData>> {
    try {
      const [settings, userProfile] = await Promise.all([
        this.settingsAccessor.getAgentSettings(new FacebookPage(pageId)),
        this.profiler.getUserProfile(
          new FacebookPage(pageId),
          new FacebookUser(pageId, userId),
        ),
      ]);

      return settings
        ? {
            ok: true,
            data: {
              page: pageId,
              user: userId,
              profile: userProfile?.data,
            },
          }
        : {
            ok: false,
            code: 404,
            reason: `Facebook page "${pageId}" not registered`,
          };
    } catch (err) {
      return err instanceof MetaApiError && err.code === 404
        ? {
            ok: false,
            code: 404,
            reason: `user "${userId}" not found or not authorized`,
          }
        : {
            ok: false,
            code: err instanceof MetaApiError ? err.code : 500,
            reason: err.message,
          };
    }
  }
}

const ServerAuthenticatorP = serviceProviderClass({
  lifetime: 'singleton',
  deps: [BotP, ProfilerP, BasicAuthenticator, AgentSettingsAccessorI],
})(FacebookServerAuthenticator);

type ServerAuthenticatorP = FacebookServerAuthenticator;

export default ServerAuthenticatorP;
