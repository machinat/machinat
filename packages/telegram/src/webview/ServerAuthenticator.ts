import { parse as parseUrl, URL } from 'url';
import type { ParsedUrlQuery } from 'querystring';
import { createHmac, createHash } from 'crypto';
import type { IncomingMessage, ServerResponse } from 'http';
import { serviceProviderClass } from '@sociably/core/service';
import type { RoutingInfo } from '@sociably/http';
import Auth, {
  ServerAuthenticator,
  VerifyResult,
  CheckDataResult,
} from '@sociably/auth';
import BasicAuthenticator from '@sociably/auth/basicAuth';
import { TELEGRAM } from '../constant.js';
import BotP from '../Bot.js';
import TelegramUser from '../User.js';
import { AgentSettingsAccessorI } from '../interface.js';
import type TelegramApiError from '../Error.js';
import type { RawChat, RawUser } from '../types.js';
import { getAuthContextDetails } from './utils.js';
import renderLoginPage from './renderLoginPage.js';
import {
  REDIRECT_QUERY,
  CHAT_ID_QUERY,
  LOGIN_PARAMETERS,
  BOT_ID_QUERY,
} from './constant.js';
import type { TelegramAuthContext, TelegramAuthData } from './types.js';

const verifyTelegramAuthQuery = (botToken: string, query: ParsedUrlQuery) => {
  const { hash } = query;
  if (typeof hash !== 'string') {
    return { ok: false as const, code: 400, reason: 'invalid auth params' };
  }

  let checkingStr = '';
  for (const paramKey of LOGIN_PARAMETERS) {
    if (query[paramKey]) {
      if (checkingStr.length !== 0) {
        checkingStr += '\n';
      }
      checkingStr += `${paramKey}=${query[paramKey]}`;
    }
  }

  const secretKey = createHash('sha256').update(botToken).digest();
  const paramsHash = createHmac('sha256', secretKey)
    .update(checkingStr)
    .digest('hex');

  if (paramsHash !== hash) {
    return { ok: false as const, code: 401, reason: 'invalid auth signature' };
  }

  if (Number(query.auth_date as string) < Date.now() / 1000 - 20) {
    return { ok: false as const, code: 401, reason: 'login expired' };
  }

  return {
    ok: true as const,
    userData: {
      id: Number(query.id),
      username: query.username,
      first_name: query.first_name,
      last_name: query.last_name,
      language_code: undefined,
    } as RawUser,
  };
};

type ServerAuthenticatorOptions = {
  appName?: string;
  appIconUrl?: string;
};

/**
 * @category Provider
 */
export class TelegramServerAuthenticator
  implements ServerAuthenticator<never, TelegramAuthData, TelegramAuthContext>
{
  bot: BotP;
  settingsAccessor: AgentSettingsAccessorI;
  operator: Auth.HttpOperator;
  appName: undefined | string;
  appIconUrl: undefined | string;

  platform = TELEGRAM;

  constructor(
    bot: BotP,
    settingsAccessor: AgentSettingsAccessorI,
    operator: Auth.HttpOperator,
    { appName, appIconUrl }: ServerAuthenticatorOptions = {}
  ) {
    this.bot = bot;
    this.settingsAccessor = settingsAccessor;
    this.operator = operator;
    this.appName = appName;
    this.appIconUrl = appIconUrl;
  }

  getAuthUrl(
    botId: number,
    chatId?: number | string,
    redirectUrl?: string
  ): string {
    const url = new URL(this.operator.getAuthUrl(TELEGRAM));
    url.searchParams.set(BOT_ID_QUERY, botId.toString());

    if (chatId) {
      url.searchParams.set(CHAT_ID_QUERY, chatId.toString());
    }
    if (redirectUrl) {
      url.searchParams.set(REDIRECT_QUERY, redirectUrl);
    }

    return url.href;
  }

  async delegateAuthRequest(
    req: IncomingMessage,
    res: ServerResponse,
    { trailingPath }: RoutingInfo
  ): Promise<void> {
    const { query } = parseUrl(req.url || '', true);

    if (trailingPath === '') {
      await this._handleTelegramAuthCallback(res, query);
    } else if (trailingPath === 'login') {
      await this._handleLoginPage(res, query);
    } else {
      res.writeHead(404);
      res.end();
    }
  }

  // eslint-disable-next-line class-methods-use-this
  async verifyCredential(): Promise<VerifyResult<never>> {
    return {
      ok: false,
      code: 403,
      reason: 'should initiate st server side only',
    };
  }

  async verifyRefreshment(
    data: TelegramAuthData
  ): Promise<VerifyResult<TelegramAuthData>> {
    const { chat, user, botId } = data;

    const [agentSettingsResult, chatMemberResult] = await Promise.all([
      this._verifyAgentSettings(botId),
      chat ? this._verifyChatMember(botId, chat.id, user.id) : null,
    ]);

    if (!agentSettingsResult.ok || (chatMemberResult && !chatMemberResult.ok)) {
      return [agentSettingsResult, chatMemberResult].find(
        (result) => result && !result.ok
      ) as VerifyResult<TelegramAuthData>;
    }

    return { ok: true, data };
  }

  // eslint-disable-next-line class-methods-use-this
  checkAuthData(data: TelegramAuthData): CheckDataResult<TelegramAuthContext> {
    return {
      ok: true,
      contextDetails: getAuthContextDetails(data),
    };
  }

  private async _handleTelegramAuthCallback(
    res: ServerResponse,
    query: ParsedUrlQuery
  ): Promise<void> {
    const targetsResult = await this._getLoginTargetsFormQuery(query);
    if (!targetsResult.ok) {
      const { code, reason, redirectUrl } = targetsResult;
      this._redirectError(res, code, reason, redirectUrl);
      return;
    }
    const { botId, agentSettings, chatId, redirectUrl } = targetsResult;

    const authQueryResult = verifyTelegramAuthQuery(
      agentSettings.botToken,
      query
    );
    if (!authQueryResult.ok) {
      const { code, reason } = authQueryResult;
      await this._redirectError(res, code, reason, redirectUrl);
      return;
    }
    const { userData } = authQueryResult;

    let chatData: undefined | RawChat;
    if (chatId) {
      const chatMemberResult = await this._verifyChatMember(
        botId,
        chatId,
        userData.id
      );

      if (!chatMemberResult.ok) {
        const { code, reason } = chatMemberResult;
        await this._redirectError(res, code, reason, redirectUrl);
        return;
      }

      const { chat, user } = chatMemberResult;
      userData.language_code = user.language_code;
      chatData = {
        type: chat.type,
        id: chat.id,
        title: chat.title,
        username: chat.username,
        description: chat.description,
      };
    }

    await this.operator.issueAuth<TelegramAuthData>(res, TELEGRAM, {
      botId,
      botName: agentSettings.botName,
      chat: chatData,
      user: userData,
      photo: query.photo_url as string | undefined,
    });
    this.operator.redirect(res, redirectUrl, { assertInternal: true });
  }

  private async _handleLoginPage(res: ServerResponse, query: ParsedUrlQuery) {
    const targetsResult = await this._getLoginTargetsFormQuery(query);
    if (!targetsResult.ok) {
      const { code, reason, redirectUrl } = targetsResult;
      this._redirectError(res, code, reason, redirectUrl);
      return;
    }
    const { botId, agentSettings, chatId, redirectUrl } = targetsResult;

    res.writeHead(200, { 'Content-Type': 'text/html' });
    res.end(
      renderLoginPage({
        botName: agentSettings.botName,
        appName: this.appName,
        appIconUrl: this.appIconUrl,
        callbackUrl: this.getAuthUrl(botId, chatId, redirectUrl),
      })
    );
  }

  private async _getLoginTargetsFormQuery(query: ParsedUrlQuery) {
    const botIdQuery = query[BOT_ID_QUERY];
    const botId =
      typeof botIdQuery === 'string' ? parseInt(botIdQuery, 10) : undefined;

    const redirectUrlQuery = query[REDIRECT_QUERY];
    const redirectUrl = Array.isArray(redirectUrlQuery)
      ? redirectUrlQuery[0]
      : redirectUrlQuery;

    const chatIdQuery = query[CHAT_ID_QUERY];
    const chatId =
      typeof chatIdQuery !== 'string'
        ? undefined
        : chatIdQuery.startsWith('@')
        ? chatIdQuery
        : parseInt(chatIdQuery, 10);

    if (!botId || (chatIdQuery && !chatId)) {
      return {
        ok: false as const,
        redirectUrl,
        code: 400,
        reason: !botId
          ? `invalid bot id "${botId}"`
          : `invalid chat id "${chatId}"`,
      };
    }

    const settingsResult = await this._verifyAgentSettings(botId);
    if (!settingsResult.ok) {
      return { ...settingsResult, redirectUrl };
    }

    return {
      ok: true as const,
      botId,
      redirectUrl,
      chatId,
      agentSettings: settingsResult.agentSettings,
    };
  }

  private async _verifyAgentSettings(botId: number) {
    const agentSettings = await this.settingsAccessor.getAgentSettings(
      new TelegramUser(botId, true)
    );
    if (!agentSettings) {
      return {
        ok: false as const,
        code: 404,
        reason: `bot "${botId}" not registered`,
      };
    }
    return { ok: true as const, agentSettings };
  }

  private async _verifyChatMember(
    botId: number,
    chatId: number | string,
    userId: number
  ) {
    const agent = new TelegramUser(botId, true);
    try {
      const [chatMember, chatData] = await Promise.all([
        this.bot.requestApi({
          agent,
          method: 'getChatMember',
          params: { chat_id: chatId, user_id: userId },
        }),
        this.bot.requestApi({
          agent,
          method: 'getChat',
          params: { chat_id: chatId },
        }),
      ]);

      if (
        chatMember.status === 'left' ||
        chatMember.status === 'kicked' ||
        (chatMember.status === 'restricted' && chatMember.is_member === false)
      ) {
        return {
          ok: false as const,
          code: 401,
          reason: 'user is unauthorized to chat',
        };
      }

      return {
        ok: true as const,
        chat: chatData as RawChat,
        user: chatMember.user as RawUser,
      };
    } catch (err) {
      const { code, description }: TelegramApiError = err;
      return { ok: false as const, code, reason: description };
    }
  }

  private async _redirectError(
    res: ServerResponse,
    code: number,
    message: string,
    redirectUrl?: string
  ) {
    await this.operator.issueError(res, TELEGRAM, code, message);
    this.operator.redirect(res, redirectUrl, { assertInternal: true });
  }
}

const ServerAuthenticatorP = serviceProviderClass({
  lifetime: 'singleton',
  deps: [
    BotP,
    AgentSettingsAccessorI,
    Auth.HttpOperator,
    { require: BasicAuthenticator, optional: true },
  ],
  factory: (bot, settingsAccessor, operator, basicAuthenticator) => {
    return new TelegramServerAuthenticator(bot, settingsAccessor, operator, {
      appName: basicAuthenticator?.appName,
      appIconUrl: basicAuthenticator?.appIconUrl,
    });
  },
})(TelegramServerAuthenticator);

type ServerAuthenticatorP = TelegramServerAuthenticator;

export default ServerAuthenticatorP;
