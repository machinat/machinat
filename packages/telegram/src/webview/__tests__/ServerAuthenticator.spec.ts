import type { IncomingMessage } from 'http';
import moxy from '@moxyjs/moxy';
import type { AuthHttpOperator } from '@machinat/auth';
import type { TelegramBot } from '../../Bot';
import { TelegramServerAuthenticator } from '../ServerAuthenticator';
import TelegramChat from '../../Chat';
import TelegramUser from '../../User';
import TelegramApiError from '../../Error';

const createReq = ({ url }) =>
  moxy<IncomingMessage>({
    method: 'GET',
    headers: {},
    url,
  } as never);

const res = {};

const httpOperator = moxy<AuthHttpOperator>({
  async issueAuth() {},
  async issueError() {},
  redirect() {
    return '';
  },
} as never);

const bot = moxy<TelegramBot>({
  id: 12345,
  token: '12345:_BOT_TOKEN_',
  makeApiCall() {
    throw new Error();
  },
} as never);

const authenticator = new TelegramServerAuthenticator(bot, httpOperator);

beforeEach(() => {
  bot.mock.reset();
  httpOperator.mock.reset();
});

const MockDate = moxy(Date);
MockDate.now.mock.fake(() => 1601136786000);
const _Date = Date;

beforeAll(() => {
  global.Date = MockDate;
});

afterAll(() => {
  global.Date = _Date;
});

describe('.delegateAuthRequest()', () => {
  const telegramLoginSearch = {
    id: '12345',
    auth_date: '1601136776',
    first_name: 'John',
    last_name: 'Doe',
    username: 'johndoe',
    photo_url: 'https://...',
    hash: 'e0aeac2cb5f34b930e6d83d16eecc5df1483c61eedf4ad07f3f43e55d6dd195d',
  };

  it('receive login request and redirct user to webview', async () => {
    const search = new URLSearchParams(telegramLoginSearch);

    const req = createReq({ url: `/auth/telegram/login?${search}` });
    await authenticator.delegateAuthRequest(req, res as never);

    expect(httpOperator.redirect.mock).toHaveBeenCalledTimes(1);
    expect(httpOperator.redirect.mock).toHaveBeenCalledWith(res, undefined, {
      assertInternal: true,
    });

    expect(httpOperator.issueError.mock).not.toHaveBeenCalled();
    expect(httpOperator.issueAuth.mock).toHaveBeenCalledTimes(1);
    expect(httpOperator.issueAuth.mock).toHaveBeenCalledWith(res, 'telegram', {
      bot: 12345,
      chat: undefined,
      user: {
        id: 12345,
        first_name: 'John',
        last_name: 'Doe',
        username: 'johndoe',
      },
      photo: 'https://...',
    });
  });

  it('verify user is a chat member if telegramChat query param given', async () => {
    const search = new URLSearchParams({
      ...telegramLoginSearch,
      telegramChat: '23456',
    });

    const req = createReq({ url: `/auth/telegram/login?${search}` });

    // getChatMember
    bot.makeApiCall.mock.fakeOnce(async () => ({
      user: {
        id: 12345,
        first_name: 'John',
        last_name: 'Doe',
        username: 'johndoe',
        language_code: 'en-US',
      },
      status: 'member',
    }));

    // getChat
    bot.makeApiCall.mock.fakeOnce(async () => ({
      id: 23456,
      type: 'group',
      title: 'Does',
    }));

    await authenticator.delegateAuthRequest(req, res as never);

    expect(bot.makeApiCall.mock).toHaveBeenCalledTimes(2);
    expect(bot.makeApiCall.mock).toHaveBeenNthCalledWith(1, 'getChatMember', {
      user_id: 12345,
      chat_id: 23456,
    });
    expect(bot.makeApiCall.mock).toHaveBeenNthCalledWith(2, 'getChat', {
      chat_id: 23456,
    });

    expect(httpOperator.issueAuth.mock).toHaveBeenCalledTimes(1);
    expect(httpOperator.issueAuth.mock).toHaveBeenCalledWith(res, 'telegram', {
      bot: 12345,
      chat: {
        type: 'group',
        id: 23456,
        title: 'Does',
      },
      user: {
        id: 12345,
        first_name: 'John',
        last_name: 'Doe',
        username: 'johndoe',
        language_code: 'en-US',
      },
      photo: 'https://...',
    });
  });

  it('fail if user is not a member of telegramChat', async () => {
    const search = new URLSearchParams({
      ...telegramLoginSearch,
      telegramChat: '23456',
    });

    const req = createReq({ url: `/auth/telegram/login?${search}` });

    bot.makeApiCall.mock.fake(async () => {
      throw new TelegramApiError({
        ok: false,
        error_code: 400,
        description: 'Bad Request: user not found',
      });
    });

    await authenticator.delegateAuthRequest(req, res as never);

    expect(bot.makeApiCall.mock).toHaveBeenCalledTimes(1);

    expect(httpOperator.issueError.mock).toHaveBeenCalledTimes(1);
    expect(httpOperator.issueError.mock).toHaveBeenCalledWith(
      res,
      'telegram',
      400,
      expect.any(String)
    );
    expect(httpOperator.issueError.mock.calls[0].args[3]).toMatchInlineSnapshot(
      `"Bad Request: user not found"`
    );
  });

  it('redirect to `redirectUrl` query param if specified', async () => {
    const search = new URLSearchParams({
      ...telegramLoginSearch,
      redirectUrl: '/webview/hello_world.html',
    });

    const req = createReq({ url: `/auth/telegram/login?${search}` });
    await authenticator.delegateAuthRequest(req, res as never);

    expect(httpOperator.redirect.mock).toHaveBeenCalledWith(
      res,
      '/webview/hello_world.html',
      { assertInternal: true }
    );

    expect(httpOperator.issueError.mock).not.toHaveBeenCalled();
    expect(httpOperator.issueAuth.mock).toHaveBeenCalledTimes(1);
  });

  it('issue error if auth_date expired (20 second)', async () => {
    const search = new URLSearchParams({
      ...telegramLoginSearch,
      auth_date: '1601136765',
      hash: '6fe5c0198e70297fffec814bf443afef5d3f6587d722399816d808e61217e571',
    });

    const req = createReq({ url: `/auth/telegram/login?${search}` });
    await authenticator.delegateAuthRequest(req, res as never);

    expect(httpOperator.redirect.mock).toHaveBeenCalledWith(res, undefined, {
      assertInternal: true,
    });

    expect(httpOperator.issueAuth.mock).not.toHaveBeenCalled();
    expect(httpOperator.issueError.mock).toHaveBeenCalledTimes(1);
    expect(httpOperator.issueError.mock).toHaveBeenCalledWith(
      res,
      'telegram',
      401,
      expect.any(String)
    );
    expect(httpOperator.issueError.mock.calls[0].args[3]).toMatchInlineSnapshot(
      `"login expired"`
    );
  });

  it('issue error if hash invalid', async () => {
    const search = new URLSearchParams({
      ...telegramLoginSearch,
      hash: '_INVALID_HASH_',
    });

    const req = createReq({
      url: `/auth/telegram?${search}`,
    });
    await authenticator.delegateAuthRequest(req, res as never);

    expect(httpOperator.redirect.mock).toHaveBeenCalledWith(res, undefined, {
      assertInternal: true,
    });

    expect(httpOperator.issueAuth.mock).not.toHaveBeenCalled();
    expect(httpOperator.issueError.mock).toHaveBeenCalledTimes(1);
    expect(httpOperator.issueError.mock).toHaveBeenCalledWith(
      res,
      'telegram',
      401,
      expect.any(String)
    );
    expect(httpOperator.issueError.mock.calls[0].args[3]).toMatchInlineSnapshot(
      `"invalid login signature"`
    );
  });
});

test('.verifyCredential() simply return not ok', async () => {
  await expect(authenticator.verifyCredential()).resolves
    .toMatchInlineSnapshot(`
          Object {
            "code": 403,
            "ok": false,
            "reason": "should initiate st server side only",
          }
        `);
});

describe('.verifyRefreshment()', () => {
  test('return ok and original data', async () => {
    const authData = {
      bot: 12345,
      chat: {
        type: 'private' as const,
        id: 67890,
      },
      user: {
        id: 67890,
        first_name: 'Jojo',
        last_name: 'Doe',
        username: 'jojodoe',
      },
      photo: undefined,
    };
    await expect(authenticator.verifyRefreshment(authData)).resolves.toEqual({
      ok: true,
      data: authData,
    });
  });

  it('return fail if botId not match', async () => {
    await expect(
      authenticator.verifyRefreshment({
        bot: 55555,
        chat: undefined,
        user: {
          id: 67890,
          first_name: 'Jojo',
          last_name: 'Doe',
          username: 'jojodoe',
        },
        photo: undefined,
      })
    ).resolves.toMatchInlineSnapshot(`
            Object {
              "code": 400,
              "ok": false,
              "reason": "bot not match",
            }
          `);
  });
});

test('.checkAuthData()', () => {
  const authData = {
    bot: 12345,
    user: {
      id: 67890,
      first_name: 'Jojo',
      last_name: 'Doe',
      username: 'jojodoe',
    },
    photo: 'http://crazy.dm/stand.png',
    chat: undefined,
  };

  const expectedUser = new TelegramUser(
    67890,
    {
      id: 67890,
      is_bot: false,
      first_name: 'Jojo',
      last_name: 'Doe',
      username: 'jojodoe',
    },
    'http://crazy.dm/stand.png'
  );

  expect(authenticator.checkAuthData(authData)).toEqual({
    ok: true,
    contextDetails: {
      botId: 12345,
      user: expectedUser,
      channel: new TelegramChat(12345, {
        type: 'private',
        id: 67890,
        first_name: 'Jojo',
        last_name: 'Doe',
        username: 'jojodoe',
      }),
      photoUrl: 'http://crazy.dm/stand.png',
    },
  });

  expect(
    authenticator.checkAuthData({
      ...authData,
      chat: { type: 'group', id: 98765 },
    })
  ).toEqual({
    ok: true,
    contextDetails: {
      botId: 12345,
      user: expectedUser,
      channel: new TelegramChat(12345, { type: 'group', id: 98765 }),
      photoUrl: 'http://crazy.dm/stand.png',
    },
  });
});
