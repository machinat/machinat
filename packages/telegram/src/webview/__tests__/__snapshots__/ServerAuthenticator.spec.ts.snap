// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.delegateAuthRequest() on login route login page with specified \`appName\` & \`appIconUrl\` 1`] = `
"
<!DOCTYPE html>
<html>
<head>
<title>Telegram Webview Login</title>
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, user-scalable=no\\">
<meta charset=\\"UTF-8\\">
<style>
body {
  margin:0;
  font-family: sans-serif;
}
.flex { display: flex }
.center {
  justify-content: center;
  align-items: center;
}
.column { flex-direction: column }
.container {
  width: 100vw;
  height: 100vh;
  gap: 1rem;
}
.icons { gap: 1.5rem }
img {
  width: 5.5rem;
  height: 5.5rem;
  object-fit: cover;
  border-radius: 0.5rem;
}
h1 {
  text-align: center;
}
.x {
  position: relative;
  margin-top: 1rem;
  height: 1rem;
  width: 1rem;
}
.x:before,
.x:after {
  position: absolute;
  content: '';
  width: 100%;
  height: 1.5px;
  background-color: black;
}
.x:before { transform: rotate(45deg) }
.x:after { transform: rotate(-45deg) }
</style>
</head>
<body>

<div class=\\"flex center column container\\">
  <div class=\\"flex center icons\\">
    <img alt=\\"Telegram\\" src=\\"https://sociably.js.org/img/icon/telegram.png\\" />
    <div class=\\"x\\"></div>
    <img alt=\\"Mine Mine Mine App\\" src=\\"http://sociably.io/MyApp/icon.png\\" />
  </div>
  <h1>Mine Mine Mine App</h1>
  <div id=\\"loginButton\\">
  </div>
</div>

<script>

const authUrl = new URL(\\"https://sociably.io/MyApp/auth/telegram?botId=12345\\");
const redirectQuery =
  new URL(window.location.href)
    .searchParams
    .get('redirectUrl');

if (redirectQuery) {
  authUrl.searchParams.set('redirectUrl', redirectQuery);
}

const loginScript = document.createElement('script');
loginScript.src = 'https://telegram.org/js/telegram-widget.js?18';
loginScript.setAttribute('async', '');
loginScript.setAttribute('data-telegram-login', 'MyBot');
loginScript.setAttribute('data-size', 'large');
loginScript.setAttribute('data-radius', '20');
loginScript.setAttribute('data-auth-url', authUrl.href);
loginScript.setAttribute('data-request-access', 'write');

const button = document.getElementById('loginButton');
button.append(loginScript);

</script>
</body>
</html>
"
`;

exports[`.delegateAuthRequest() on login route render login page with Telegarm login button 1`] = `
"
<!DOCTYPE html>
<html>
<head>
<title>Telegram Webview Login</title>
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, user-scalable=no\\">
<meta charset=\\"UTF-8\\">
<style>
body {
  margin:0;
  font-family: sans-serif;
}
.flex { display: flex }
.center {
  justify-content: center;
  align-items: center;
}
.column { flex-direction: column }
.container {
  width: 100vw;
  height: 100vh;
  gap: 1rem;
}
.icons { gap: 1.5rem }
img {
  width: 5.5rem;
  height: 5.5rem;
  object-fit: cover;
  border-radius: 0.5rem;
}
h1 {
  text-align: center;
}
.x {
  position: relative;
  margin-top: 1rem;
  height: 1rem;
  width: 1rem;
}
.x:before,
.x:after {
  position: absolute;
  content: '';
  width: 100%;
  height: 1.5px;
  background-color: black;
}
.x:before { transform: rotate(45deg) }
.x:after { transform: rotate(-45deg) }
</style>
</head>
<body>

<div class=\\"flex center column container\\">
  <div class=\\"flex center icons\\">
    <img alt=\\"Telegram\\" src=\\"https://sociably.js.org/img/icon/telegram.png\\" />
  </div>
  <h1></h1>
  <div id=\\"loginButton\\">
  </div>
</div>

<script>

const authUrl = new URL(\\"https://sociably.io/MyApp/auth/telegram?botId=12345\\");
const redirectQuery =
  new URL(window.location.href)
    .searchParams
    .get('redirectUrl');

if (redirectQuery) {
  authUrl.searchParams.set('redirectUrl', redirectQuery);
}

const loginScript = document.createElement('script');
loginScript.src = 'https://telegram.org/js/telegram-widget.js?18';
loginScript.setAttribute('async', '');
loginScript.setAttribute('data-telegram-login', 'MyBot');
loginScript.setAttribute('data-size', 'large');
loginScript.setAttribute('data-radius', '20');
loginScript.setAttribute('data-auth-url', authUrl.href);
loginScript.setAttribute('data-request-access', 'write');

const button = document.getElementById('loginButton');
button.append(loginScript);

</script>
</body>
</html>
"
`;

exports[`.delegateAuthRequest() on login route with \`chatId\` & \`redirectUrl\` specified 1`] = `
"
<!DOCTYPE html>
<html>
<head>
<title>Telegram Webview Login</title>
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, user-scalable=no\\">
<meta charset=\\"UTF-8\\">
<style>
body {
  margin:0;
  font-family: sans-serif;
}
.flex { display: flex }
.center {
  justify-content: center;
  align-items: center;
}
.column { flex-direction: column }
.container {
  width: 100vw;
  height: 100vh;
  gap: 1rem;
}
.icons { gap: 1.5rem }
img {
  width: 5.5rem;
  height: 5.5rem;
  object-fit: cover;
  border-radius: 0.5rem;
}
h1 {
  text-align: center;
}
.x {
  position: relative;
  margin-top: 1rem;
  height: 1rem;
  width: 1rem;
}
.x:before,
.x:after {
  position: absolute;
  content: '';
  width: 100%;
  height: 1.5px;
  background-color: black;
}
.x:before { transform: rotate(45deg) }
.x:after { transform: rotate(-45deg) }
</style>
</head>
<body>

<div class=\\"flex center column container\\">
  <div class=\\"flex center icons\\">
    <img alt=\\"Telegram\\" src=\\"https://sociably.js.org/img/icon/telegram.png\\" />
  </div>
  <h1></h1>
  <div id=\\"loginButton\\">
  </div>
</div>

<script>

const authUrl = new URL(\\"https://sociably.io/MyApp/auth/telegram?botId=12345&chatId=67890&redirectUrl=%2Fwebview%2Fhello_world\\");
const redirectQuery =
  new URL(window.location.href)
    .searchParams
    .get('redirectUrl');

if (redirectQuery) {
  authUrl.searchParams.set('redirectUrl', redirectQuery);
}

const loginScript = document.createElement('script');
loginScript.src = 'https://telegram.org/js/telegram-widget.js?18';
loginScript.setAttribute('async', '');
loginScript.setAttribute('data-telegram-login', 'MyBot');
loginScript.setAttribute('data-size', 'large');
loginScript.setAttribute('data-radius', '20');
loginScript.setAttribute('data-auth-url', authUrl.href);
loginScript.setAttribute('data-request-access', 'write');

const button = document.getElementById('loginButton');
button.append(loginScript);

</script>
</body>
</html>
"
`;
